### PR TITLE
Change `rustc_codegen_ssa`'s `atomic_cmpxchg` interface to return a pair of values

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Doesn't change much, but a little nicer that way.

### CODE REVIEW

File: compiler/rustc_codegen_gcc/src/builder.rs [hunk 1]
```diff
@@ -1310,20 +1310,7 @@ impl<'a, 'gcc, 'tcx> BuilderMethods<'a, 'tcx> for Builder<'a, 'gcc, 'tcx> {
...
-        if let Some(struct_type) = value_type.is_struct() {
-            self.store(success, result.access_field(None, struct_type.get_field(1)).get_address(None), align);
-            // NOTE: since success contains the call to the intrinsic, it must be stored before
-            // expected so that we store expected after the call.
-            self.store(expected.to_rvalue(), result.access_field(None, struct_type.get_field(0)).get_address(None), align);
-        }
-        // TODO(antoyo): handle when value is not a struct.
-
-        result.to_rvalue()
+        ( expected.to_rvalue(), success )
```
Comment (GuillaumeGomez) [inline compiler/rustc_codegen_gcc/src/builder.rs:1313]:
```suggestion
        (expected.to_rvalue(), success)
```

**Comment by Urgau** [comment]
> Seems like it's failing because of cg_gcc tests (added back in #118463). Not sure why though.

@WaffleLapkin This has been fixed in https://github.com/rust-lang/rust/pull/118706, it was a general issue with PR CI, nothing about your PR. Rebasing should be enough to fix this build issue.

**Comment by WaffleLapkin** [comment]
@rustbot ready

Indeed, rebase seems to have fixed ci. Thanks, @Urgau!

**Comment by cjgillot** [comment]
r=me with @antoyo's green light on the cg_gcc change.

**Comment by antoyo** [comment]
I see this removed some stuff like the alignment that might still be necessary.

@WaffleLapkin: Did you run all the rustc_codegen_gcc's test suite locally (the `test.sh` script)?

**Comment by WaffleLapkin** [comment]
@antoyo I don't think alignment is necessary, it was only used for the stores into the tuple temp. I did not run the tests and looking at it, they are quite cumbersome to setup, so I don't think I'll do it.

If you want to run tests yourself I suggest doing `git fetch origin 0e4ccff807cce2f78ada9e3dda868bcdb419564d && git checkout 0e4ccff807cce2f78ada9e3dda868bcdb419564d`. GitHub is setup in a way that you can just do that.

**Comment by antoyo** [comment]
Ok. I'll run the whole test suite locally just to make sure.

**Comment by antoyo** [comment]
There's this libcore test that is failing: `atomic::bool_`.
I'll investigate.

**Comment by antoyo** [comment]
This might also caused by this thing that was noted and changed:
> since success contains the call to the intrinsic, it must be stored before expected so that we store expected after the call.

**Comment by antoyo** [comment]
> This might also caused by this thing that was noted and changed:
> 
> > since success contains the call to the intrinsic, it must be stored before expected so that we store expected after the call.

Yeah, that was it. The following code makes that test pass:
```rust
    fn atomic_cmpxchg(&mut self, dst: RValue<'gcc>, cmp: RValue<'gcc>, src: RValue<'gcc>, order: AtomicOrdering, failure_order: AtomicOrdering, weak: bool) -> (RValue<'gcc>, RValue<'gcc>) {
        let expected = self.current_func().new_local(None, cmp.get_type(), "expected");
        self.llbb().add_assignment(None, expected, cmp);
        // NOTE: gcc doesn't support a failure memory model that is stronger than the success
        // memory model.
        let order =
            if failure_order as i32 > order as i32 {
                failure_order
            }
            else {
                order
            };
        let success = self.compare_exchange(dst, expected, src, order, failure_order, weak);

        let success_var = self.current_func().new_local(None, self.bool_type, "success");
        self.llbb().add_assignment(None, success_var, success);

        (expected.to_rvalue(), success_var.to_rvalue())
    }
```

I haven't ran the rest of the test suite yet.

**Pull request merged**